File: wts-client-java/src/net/sf/wts/client/WtsClient.java
package net.sf.wts.client;

import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.rmi.RemoteException;

import net.sf.wts.client.util.AxisClient;
import net.sf.wts.client.util.Encrypt;

import org.apache.commons.io.FileUtils;

public class WtsClient implements IWtsClient
{
    private String url_;
    private AxisClient axisService = new AxisClient();
    
    private int retryOnRemoteException;
    
    public WtsClient(String url, int retryOnRemoteException)
    {
        url_ = url;
        this.retryOnRemoteException = retryOnRemoteException;
    }
    
    public WtsClient(String url)
    {
    	this(url, 1);
    }
    
    public String getChallenge(String userName) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			return getChallengeImpl(userName);
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    	
    	return null;
    }
    
    private String getChallengeImpl(String userName) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "GetChallenge");
        
        axisService.addParameter(userName);
        
        String challenge = "";
        
        challenge = axisService.callAndGetStringResult();
        
        return challenge;
    }
    
    public String login(String userName, String challenge, String password, String serviceName) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			return loginImpl(userName, challenge, password, serviceName);
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    	
    	return null;
    }
    
    private String loginImpl(String userName, String challenge, String password, String serviceName) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "Login");
        
        axisService.addParameter(userName);
        axisService.addParameter(challenge);
        axisService.addParameter(Encrypt.encryptMD5(Encrypt.encryptMD5(password)+challenge));
        axisService.addParameter(serviceName);
        
        String sessionTicket = "";
        
        sessionTicket = axisService.callAndGetStringResult();
        
        return sessionTicket;
    }
    
    public void upload(String sessionTicket, String serviceName, String fileName, byte[] file) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			uploadImpl(sessionTicket, serviceName, fileName, file);
    			break;
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    }
    
    private void uploadImpl(String sessionTicket, String serviceName, String fileName, byte[] file) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "Upload");
        
        axisService.addParameter(sessionTicket);
        axisService.addParameter(serviceName);
        axisService.addParameter(fileName);
        axisService.addParameter(file);
        
        try 
        {
            axisService.call();
        } 
        catch (RemoteException e) 
        {
            e.printStackTrace();
        }
    }
    
    public byte[] download(String sessionTicket, String serviceName, String fileName) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			return downloadImpl(sessionTicket, serviceName, fileName);
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    	
    	return null;
    }
    
    private byte[] downloadImpl(String sessionTicket, String serviceName, String fileName) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "Download");
        
        axisService.addParameter(sessionTicket);
        axisService.addParameter(serviceName);
        axisService.addParameter(fileName);
        
        byte[] array = null;
        try 
        {
            array = axisService.callAndGetByteArrayResult();
        } 
        catch (RemoteException e) 
        {
            e.printStackTrace();
        }
        
        return array;
    }
    
    public void download(String sessionTicket, String serviceName, String fileName, File toWrite) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			downloadImpl(sessionTicket, serviceName, fileName, toWrite);
    			break;
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    }
    
    private void downloadImpl(String sessionTicket, String serviceName, String fileName, File toWrite) throws RemoteException, MalformedURLException
    {
        byte[] array = download(sessionTicket, serviceName, fileName);
        if(array!=null)
        {
            try 
            {
                FileUtils.writeByteArrayToFile(toWrite, array);
            } 
            catch (IOException e) 
            {
                e.printStackTrace();
                throw new RuntimeException(e);
            }
        }
    }
    
    public void upload(String sessionTicket, String serviceName, String fileName, File localLocation) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			uploadImpl(sessionTicket, serviceName, fileName, localLocation);
    			break;
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    }
    
    private void uploadImpl(String sessionTicket, String serviceName, String fileName, File localLocation) throws RemoteException, MalformedURLException
    {
        byte[] array = null;
        try 
        {
            array = FileUtils.readFileToByteArray(localLocation);;
        } 
        catch (IOException e) 
        {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
        
        upload(sessionTicket, serviceName, fileName, array);
    }
    
    public byte[] monitorLogFile(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			return monitorLogFileImpl(sessionTicket, serviceName);
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    	
    	return null;
    }
    
    private byte[] monitorLogFileImpl(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "MonitorLogFile");
        
        axisService.addParameter(sessionTicket);
        axisService.addParameter(serviceName);

        byte[] array = null;
        try 
        {
            array = axisService.callAndGetByteArrayResult();
        } 
        catch (RemoteException e) 
        {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
        
        return array;
    }
    
    public byte[] monitorLogTail(String sessionTicket, String serviceName, int numberOfLines) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			return monitorLogTail(sessionTicket, serviceName, numberOfLines);
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    	
    	return null;
    }
    
    private byte[] monitorLogTailImpl(String sessionTicket, String serviceName, int numberOfLines) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "MonitorLogTail");
        
        axisService.addParameter(sessionTicket);
        axisService.addParameter(serviceName);
        axisService.addParameter(numberOfLines);

        byte[] array = null;
        try 
        {
            array = axisService.callAndGetByteArrayResult();
        } 
        catch (RemoteException e) 
        {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
        
        return array;
    }
    
    public String monitorStatus(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			return monitorStatusImpl(sessionTicket, serviceName);
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    	
    	return null;
    }
    
    private String monitorStatusImpl(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "MonitorStatus");
        
        axisService.addParameter(sessionTicket);
        axisService.addParameter(serviceName);
        
        String status = "";
        
        try 
        {
            status = axisService.callAndGetStringResult();
        } 
        catch (RemoteException e) 
        {
            e.printStackTrace();
            throw new RuntimeException(e);
        }

        return status;
    }
    
    public void start(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			startImpl(sessionTicket, serviceName);
    			break;
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    }
    
    private void startImpl(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "Start");
        
        axisService.addParameter(sessionTicket);
        axisService.addParameter(serviceName);
        
        try 
        {
            axisService.call();
        } 
        catch (RemoteException e)
        {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }
    
    public void closeSession(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			closeSessionImpl(sessionTicket, serviceName);
    			break;
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    }
    
    private void closeSessionImpl(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "CloseSession");
        
        axisService.addParameter(sessionTicket);
        axisService.addParameter(serviceName);
        
        try 
        {
            axisService.call();
        } 
        catch (RemoteException e)
        {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }
    
    public void stop(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException {
    	for(int i = 0; i<retryOnRemoteException+1; ++i) {
    		try {
    			stopImpl(sessionTicket, serviceName);
    			break;
    		} catch(RemoteException re) {
    			if(i==retryOnRemoteException) {
    				throw re;
    			}
    		}
    	}
    }
    
    private void stopImpl(String sessionTicket, String serviceName) throws RemoteException, MalformedURLException
    {
        axisService.removeParameters();
        axisService.setServiceUrl(url_, "Stop");
        
        axisService.addParameter(sessionTicket);
        axisService.addParameter(serviceName);
        
        try 
        {
            axisService.call();
        } 
        catch (RemoteException e)
        {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    public String getUrl() 
    {
        return url_;
    }
}

File: wts-client/net/sf/wts/client/WtsClient.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using net.sf.wts.client.proxyFiles;
using net.sf.wts.client.proxyFiles.GetChallenge;
using net.sf.wts.client.proxyFiles.Login;
using System.Security.Cryptography;
using net.sf.wts.client.proxyFiles.Upload;
using net.sf.wts.client.util;
using net.sf.wts.client.proxyFiles.Download;
using System.IO;
using net.sf.wts.client.proxyFiles.Start;
using net.sf.wts.client.proxyFiles.MonitorLogFile;
using net.sf.wts.client.proxyFiles.MonitorLogTail;
using net.sf.wts.client.proxyFiles.MonitorStatus;
using net.sf.wts.client.proxyFiles.CloseSession;
using net.sf.wts.client.proxyFiles.Stop;

namespace net.sf.wts.client
{
    public class WtsClient
    {
        private string url_;
        private GetChallengeService getChallengeService_ = null;
        private LoginService loginService_ = null;
        private UploadService UploadService_ = null;
        private DownloadService DownloadService_ = null;
        private MonitorLogFileService MonitorLogFileService_ = null;
        private MonitorLogTailService MonitorLogTailService_ = null;
        private MonitorStatusService MonitorStatusService_ = null;
        private StartService StartService_ = null;
        private CloseSessionService CloseSessionService_ = null;
        private StopService StopService_ = null;

        public WtsClient(string url)
        {
            url_ = url;
        }

        public string GetChallenge(string userName)
        {
            if (getChallengeService_ == null)
            {
                getChallengeService_ = new GetChallengeService();
                getChallengeService_.Url = url_+"GetChallenge";
            }

            return getChallengeService_.exec(userName);
        }
     
        public string Login(string userName, string challenge, string password, string serviceName)
        {
            if (loginService_ == null)
            {
                loginService_ = new LoginService();
                loginService_.Url = url_ + "Login";
            }

            string encrypted = EncodePassword(EncodePassword(password) + challenge);

            return loginService_.exec(userName, challenge, encrypted, serviceName);
        }

        public void Upload(string sessionTicket, string serviceName, string fileName, byte[] file)
        {
            if (UploadService_ == null)
            {
                UploadService_ = new UploadService();
                UploadService_.Url = url_ + "Upload";
            }

            UploadService_.exec(sessionTicket, serviceName, fileName, file);
        }

        public void Upload(string sessionTicket, string serviceName, string fileName, string fileLocalLocation)
        {
            if (UploadService_ == null)
            {
                UploadService_ = new UploadService();
                UploadService_.Url = url_ + "Upload";
            }

            byte[] array = FileUtils.ReadFileToArray(fileLocalLocation);

            UploadService_.exec(sessionTicket, serviceName, fileName, array);
        }

        public byte[] Download(string sessionTicket, string serviceName, string fileName)
        {
            if (DownloadService_ == null)
            {
                DownloadService_ = new DownloadService();
                DownloadService_.Url = url_ + "Download";
            }

            return DownloadService_.exec(sessionTicket, serviceName, fileName);
        }

        public void Download(string sessionTicket, string serviceName, string fileName, string fileToWrite)
        {
            byte[] array = Download(sessionTicket, serviceName, fileName);
            if (array != null)
            {
                FileUtils.writeByteArrayToFile(fileToWrite, array);
            }
        }

        public byte[] MonitorLogFile(String sessionTicket, String serviceName)
        {
            if (MonitorLogFileService_ == null)
            {
                MonitorLogFileService_ = new MonitorLogFileService();
                MonitorLogFileService_.Url = url_ + "MonitorLogFile";
            }

            return MonitorLogFileService_.exec(sessionTicket, serviceName);
        }
        
        public byte[] MonitorLogTail(String sessionTicket, String serviceName, int numberOfLines)
        {
            if (MonitorLogTailService_ == null)
            {
                MonitorLogTailService_ = new MonitorLogTailService();
                MonitorLogTailService_.Url = url_ + "MonitorLogTail";
            }

            return MonitorLogTailService_.exec(sessionTicket, serviceName, numberOfLines);
        }

        public String MonitorStatus(String sessionTicket, String serviceName)
        {
            if (MonitorStatusService_ == null)
            {
                MonitorStatusService_ = new MonitorStatusService();
                MonitorStatusService_.Url = url_ + "MonitorStatus";
            }

            return MonitorStatusService_.exec(sessionTicket, serviceName);
        }

        public void Start(string sessionTicket, string serviceName)
        {
            if (StartService_ == null)
            {
                StartService_ = new StartService();
                StartService_.Url = url_ + "Start";
            }

            StartService_.exec(sessionTicket, serviceName);
        }

        public void CloseSession(string sessionTicket, string serviceName)
        {
            if (CloseSessionService_ == null)
            {
                CloseSessionService_ = new CloseSessionService();
                CloseSessionService_.Url = url_ + "CloseSession";
            }

            CloseSessionService_.exec(sessionTicket, serviceName);
        }

        public void Stop(string sessionTicket, string serviceName)
        {
            if (StopService_ == null)
            {
                StopService_ = new StopService();
                StopService_.Url = url_ + "Stop";
            }

            StopService_.exec(sessionTicket, serviceName);
        }

        public string getUrl()
        {
            return url_;
        }

        private string EncodePassword(string originalPassword)
        {
            byte[] originalBytes;
            byte[] encodedBytes;
            MD5 md5;

            md5 = new MD5CryptoServiceProvider();
            UTF8Encoding enc = new UTF8Encoding();

            originalBytes = enc.GetBytes(originalPassword);
            encodedBytes = md5.ComputeHash(originalBytes);

            return Convert.ToBase64String(encodedBytes);
        }

        public static void Main(string[] args)
        {
            WtsClient wtsclient = new WtsClient("http://zolder:8080/wts/services/");
            string challenge = wtsclient.GetChallenge("kdforc0");
            Console.WriteLine(challenge);
            string ticket = wtsclient.Login("kdforc0", challenge, "Vitabis1", "regadb-hiv-type");
            Console.WriteLine(ticket);
            wtsclient.Upload(ticket, "regadb-hiv-type", "nt_sequence", "C:\\jvsant1\\isolate.fasta");
            wtsclient.Start(ticket, "regadb-hiv-type");

            Thread.Sleep(5000);

            Console.WriteLine("monitor log file:");
            string logFile = System.Text.Encoding.UTF8.GetString(wtsclient.MonitorLogFile(ticket, "regadb-hiv-type"));
            Console.WriteLine(logFile);

            Console.WriteLine("monitor log tail:");
            string logTail = System.Text.Encoding.UTF8.GetString(wtsclient.MonitorLogTail(ticket, "regadb-hiv-type", 10));
            Console.WriteLine(logTail);

            Console.WriteLine("status:" + wtsclient.MonitorStatus(ticket, "regadb-hiv-type"));
            wtsclient.Download(ticket, "regadb-hiv-type", "type", "C:\\jvsant1\\type.txt");

            wtsclient.CloseSession(ticket, "regadb-hiv-type");

            Console.ReadLine();
        }
    }
}
